### PR TITLE
Add Obscuro gateway endpoints

### PR DESCRIPTION
**Pull request by zkokelj**
### Why this change is needed

As described in [Obscuro Gateway Design](https://github.com/obscuronet/go-obscuro/blob/main/design/ux/Obscuro_Gateway.md ) new endpoints are needed

### What changes were made as part of this PR

Added following endpoints:
- /join
- /authenticate
- /revoke
- /query

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/obscuronet/obscuro-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [ ] PR checks reviewed and performed

### CODE REVIEW

**Comment by tudor-malene** [comment]
> ### Why this change is needed
> Please provide a description and a link to the underlying ticket
> 
> ### What changes were made as part of this PR
> Please provide a high level list of the changes made
> 
> ### PR checks pre-merging
> Please indicate below by ticking the checkbox that you have read and performed the required [PR checks](https://github.com/obscuronet/obscuro-internal/blob/main/dev-ops-docs/dev-pr-checks.md)
> 
> * [ ]  PR checks reviewed and performed

description pls

File: tools/walletextension/api/routes.go [hunk 1]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
 		return
 	}
 }
+
+func joinRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
+	// todo (@ziga) add protection against DDOS attacks
+	_, err := userConn.ReadRequest()
+	if err != nil {
+		return
+	}
```
Comment (otherview) [inline tools/walletextension/api/routes.go:246]: Not sure of the purpose of reading the request and swallowing any error.

File: tools/walletextension/api/routes.go [hunk 2]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+
+	// create UserID and store it in the database with the private key
+	userID := crypto.Keccak256Hash(viewingPublicKeyBytes)
+	err = walletExt.Storage.AddUser(userID.Bytes(), crypto.FromECDSA(viewingPrivateKeyEcies.ExportECDSA()))
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Sprintf("failed to save user to the database: %s", err))
+		walletExt.Logger().Error(fmt.Sprintf("failed to save user to the database: %s", err))
+		return
+	}
```
Comment (otherview) [inline tools/walletextension/api/routes.go:265]: The route should request the WE to handle the VK gen. The route doesn't need to know about vk storage, etc.

File: tools/walletextension/api/routes.go [hunk 3]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	// get the text that was signed and signature
+	var reqJSONMap map[string]string
+	err = json.Unmarshal(body, &reqJSONMap)
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("error unmarshaling request to authentcate: %w", err).Error())
+		return
+	}
+
+	// get signature from the request and remove leading two bytes (0x)
```
Comment (otherview) [inline tools/walletextension/api/routes.go:290]: Clean comments, nice !

File: tools/walletextension/api/routes.go [hunk 4]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	signature, err := hex.DecodeString(reqJSONMap[common.JSONKeySignature][2:])
+	if err != nil {
+		userConn.HandleError("Internal error")
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:293]: this is not an internal error. It's a malformed query. So it's relevant to be returned to the user

File: tools/walletextension/api/routes.go [hunk 5]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	}
+
+	// check if userID corresponds to the one in the message
+	// todo: do we need userID in query param, because we get it already from the message
+	if userID != messageUserID || len(messageUserID) != 66 {
+		userConn.HandleError(fmt.Sprintf("User in submitted message (%s) does not match user provided in the request (%s) of is wrong size", messageUserID, userID))
+		return
+	}
+
+	// Check if the signature is valid
```
Comment (otherview) [inline tools/walletextension/api/routes.go:335]: This is a todo ?
Likely worth moving signature validation to the WE ?

File: tools/walletextension/api/routes.go [hunk 6]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("could not find or decode signature from client to hex: %w", err).Error())
+		return
+	}
+
+	// get message from the request
+	message, ok := reqJSONMap[common.JSONKeyMessage]
+	if !ok || message == "" {
+		userConn.HandleError("Internal error")
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:301]: this is not an internal error. It's a malformed query. So it's relevant to be returned to the user

File: tools/walletextension/api/routes.go [hunk 7]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		messageAddressHex = params[2]
+	} else {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("submitted message (%s) is not in the correct format", message).Error())
+		return
+	}
+
+	// check if userID corresponds to the one in the message
+	// todo: do we need userID in query param, because we get it already from the message
+	if userID != messageUserID || len(messageUserID) != 66 {
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:330]: magic number 66. Should be a constant with a comment

File: tools/walletextension/api/routes.go [hunk 8]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		userConn.HandleError(fmt.Sprintf("User in submitted message (%s) does not match user provided in the request (%s) of is wrong size", messageUserID, userID))
+		return
+	}
+
+	// Check if the signature is valid
+
+	// prefix the message like in the personal_sign method
+	prefixedMessage := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(message), message)
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:338]: constant

File: tools/walletextension/api/routes.go [hunk 9]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	messageHash := crypto.Keccak256([]byte(prefixedMessage))
+
+	// check the signature length
+	if len(signature) != 65 {
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:342]: magic number

File: tools/walletextension/api/routes.go [hunk 10]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		return
+	}
+}
+
+func queryRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
+	// read the request
+	_, err := userConn.ReadRequest()
+	if err != nil {
+		walletExt.Logger().Error(fmt.Errorf("error reading request: %w", err).Error())
+		return
```
Comment (otherview) [inline tools/walletextension/api/routes.go:395]: Think it's fine to return bad request reading to the user ?

File: tools/walletextension/api/utils.go [hunk 11]
```diff
@@ -37,3 +37,12 @@ func parseRequest(body []byte) (*accountmanager.RPCRequest, error) {
 		Params: params,
 	}, nil
 }
+
+func getQueryParameter(params map[string]string, selectedParameter string) (string, error) {
+	value, exists := params[selectedParameter]
+	if !exists {
+		return "", fmt.Errorf("parameter 'u' is not in the query params")
```
Comment (otherview) [inline tools/walletextension/api/utils.go:44]:
```suggestion
		return "", fmt.Errorf("parameter '%s' is not in the query params", selectedParameter)
```

File: tools/walletextension/api/routes.go [hunk 12]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	// Check if the signature is valid
+
+	// prefix the message like in the personal_sign method
+	prefixedMessage := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(message), message)
+	messageHash := crypto.Keccak256([]byte(prefixedMessage))
+
+	// check the signature length
+	if len(signature) != 65 {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("signature must be 64 bytes long, but %d bytes long signature received", len(signature)).Error())
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:344]: this is not an internal error. It's a malformed query. So it's relevant to be returned to the user

File: tools/walletextension/api/routes.go [hunk 13]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+
+func queryRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
+	// read the request
+	_, err := userConn.ReadRequest()
+	if err != nil {
+		walletExt.Logger().Error(fmt.Errorf("error reading request: %w", err).Error())
+		return
+	}
+
+	userID, err := getQueryParameter(userConn.ReadRequestParams(), "u")
```
Comment (otherview) [inline tools/walletextension/api/routes.go:398]: Suggestion: `userID` -> `hexUserID`

File: tools/walletextension/api/routes.go [hunk 14]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		walletExt.Logger().Error(fmt.Errorf("signature must be 64 bytes long, but %d bytes long signature received", len(signature)).Error())
+		return
+	}
+
+	// We transform the V from 27/28 to 0/1. This same change is made in Geth internals, for legacy reasons to be able
+	// to recover the address: https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L452-L459
+	signature[64] -= 27
+
+	// check if the message contains the same address that signed the message
+	pubKey, err := crypto.SigToPub(messageHash, signature)
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:353]: I would also extract this check in a function

File: tools/walletextension/api/routes.go [hunk 15]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		return
+	}
+
+	err = userConn.WriteResponse(msg)
+	if err != nil {
+		return
+	}
+}
+
+func revokeRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
```
Comment (otherview) [inline tools/walletextension/api/routes.go:458]: Comment please

File: tools/walletextension/api/routes.go [hunk 16]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		walletExt.Logger().Error(fmt.Errorf("error decoding string (%s), %w", userID, err).Error())
+		return
+	}
+
+	err = walletExt.Storage.DeleteUser(userIDBytes)
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("error deleting user (%s), %w", userID, err).Error())
+		return
+	}
```
Comment (otherview) [inline tools/walletextension/api/routes.go:484]: This should be done inside the `wallet_extension.go`

File: tools/walletextension/api/routes.go [hunk 17]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+}
```
Comment (otherview) [inline tools/walletextension/api/routes.go:485]: Should a success message be returned? Or does that happen by default ?

File: tools/walletextension/api/routes.go [hunk 18]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		walletExt.Logger().Error(fmt.Errorf("error getting accounts for user (%s), %w", userID, err).Error())
+		return
+	}
+
+	found := false
+	for _, account := range accounts {
+		if bytes.Equal(account.AccountAddress, addressBytes) {
+			found = true
+		}
+	}
```
Comment (otherview) [inline tools/walletextension/api/routes.go:438]: This is a method of Wallet Extension. The routing shouldn't access the storage right ?

File: tools/walletextension/api/routes.go [hunk 19]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("error decoding string (%s), %w", userID[2:], err).Error())
+		return
+	}
+	err = walletExt.Storage.AddAccount(userIDBytes, addressFromMessage.Bytes(), signature)
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("error while storing account (%s) for user (%s): %w", addressFromMessage.Hex(), userID, err).Error())
+		return
+	}
```
Comment (otherview) [inline tools/walletextension/api/routes.go:382]: This should be a method of the wallet extension.

**Review by tudor-malene — CHANGES_REQUESTED** [review]
looks good in principle.

Left a few aesthetic comments to make the code more readable, and some around error handling

File: tools/walletextension/wallet_extension.go [hunk 20]
```diff
@@ -44,7 +44,7 @@ func New(
 		hostAddr:           hostAddr,
 		userAccountManager: userAccountManager,
 		unsignedVKs:        map[gethcommon.Address]*rpc.ViewingKey{},
-		storage:            storage,
+		Storage:            storage,
```
Comment (otherview) [inline tools/walletextension/wallet_extension.go:47]: Storage should be a private prop. Only the WE should access it.

File: tools/walletextension/api/routes.go [hunk 21]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+
+func revokeRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
+	// read the request
+	_, err := userConn.ReadRequest()
+	if err != nil {
+		walletExt.Logger().Error(fmt.Errorf("error reading request: %w", err).Error())
+		return
+	}
+
+	userID, err := getQueryParameter(userConn.ReadRequestParams(), "u")
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:466]: fixed

File: tools/walletextension/api/routes.go [hunk 22]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	_, err := userConn.ReadRequest()
+	if err != nil {
+		return
+	}
+
+	// generate new key-pair
+	viewingKeyPrivate, err := crypto.GenerateKey()
+	viewingPrivateKeyEcies := ecies.ImportECDSA(viewingKeyPrivate)
+	if err != nil {
+		userConn.HandleError(fmt.Sprintf("could not generate new keypair: %s", err))
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:252]: fixed.

File: tools/walletextension/api/routes.go [hunk 23]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		return
+	}
+	viewingPublicKeyBytes := crypto.CompressPubkey(&viewingKeyPrivate.PublicKey)
+
+	// create UserID and store it in the database with the private key
+	userID := crypto.Keccak256Hash(viewingPublicKeyBytes)
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:258]: added new function

File: tools/walletextension/api/routes.go [hunk 24]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	err = walletExt.Storage.AddUser(userID.Bytes(), crypto.FromECDSA(viewingPrivateKeyEcies.ExportECDSA()))
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Sprintf("failed to save user to the database: %s", err))
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:262]: removed duplicate

File: tools/walletextension/api/routes.go [hunk 25]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("could not find message in the request: %w", err).Error())
+		return
+	}
+
+	// read userID from query params
+	userID, err := getQueryParameter(userConn.ReadRequestParams(), "u")
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("user not found in the query params: %w", err).Error())
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:310]: I agree, will change this and other similar errors to be displayed to users (if relevant to them)

File: tools/walletextension/api/routes.go (hunk 4 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go:293]: fixed

File: tools/walletextension/api/routes.go (hunk 8 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go:338]: added

File: tools/walletextension/api/routes.go (hunk 12 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go:344]: changed message for the user

File: tools/walletextension/api/routes.go (hunk 14 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go:353]: extracted

File: tools/walletextension/api/routes.go [hunk 26]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	if addressFromSignature != addressFromMessage {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("address from signature (%s) is not the same as address from message (%s)", addressFromSignature, addressFromSignature).Error())
+		return
+	}
+
+	// register the account for that viewing key
+	userIDBytes, err := hex.DecodeString(userID[2:]) // remove 0x prefix from userID
+	if err != nil {
+		userConn.HandleError("Internal error")
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:373]: fixed

File: tools/walletextension/api/routes.go [hunk 27]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+
+	userID, err := getQueryParameter(userConn.ReadRequestParams(), "u")
+	if err != nil {
+		userConn.HandleError("user ('u') not found in query parameters")
+		walletExt.Logger().Error(fmt.Errorf("user not found in the query params: %w", err).Error())
+		return
+	}
+	userIDBytes, err := hex.DecodeString(userID[2:]) // remove 0x prefix from userID
+	if err != nil {
+		userConn.HandleError("Internal error")
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:406]: Added new function

File: tools/walletextension/api/routes.go [hunk 28]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
 		return
 	}
 }
+
+func joinRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:241]: added

File: tools/walletextension/api/routes.go (hunk 1 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go:246]: logging added

File: tools/walletextension/api/routes.go (hunk 17 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go:485]: success message added

File: tools/walletextension/api/routes.go [hunk 29]
```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		return
+	}
+
+	err = userConn.WriteResponse([]byte("success!"))
+	if err != nil {
+		return
+	}
+}
+
+func queryRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:390]: added

File: tools/walletextension/api/routes.go (hunk 10 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go:395]: returning it to the user

File: tools/walletextension/api/utils.go (hunk 11 above)
Comment (zkokelj) [reply tools/walletextension/api/utils.go:44]: fixed, thanks

File: tools/walletextension/api/routes.go (hunk 13 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go:398]: changed

File: tools/walletextension/api/routes.go (hunk 15 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go:458]: comment added

File: tools/walletextension/api/routes.go (hunk 2 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go]: moved to WE

File: tools/walletextension/api/routes.go (hunk 18 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go]: moved to WE

File: tools/walletextension/api/routes.go (hunk 19 above)
Comment (zkokelj) [reply tools/walletextension/api/routes.go]: moved to WE

File: tools/walletextension/wallet_extension.go (hunk 20 above)
Comment (zkokelj) [reply tools/walletextension/wallet_extension.go]: storate is private again, moved all the code accessing it to WE

File: tools/walletextension/wallet_extension.go [hunk 30]
```diff
@@ -159,6 +162,165 @@ func (w *WalletExtension) SubmitViewingKey(address gethcommon.Address, signature
...
+		return err
+	}
+
+	// check if userID corresponds to the one in the message and check if the length of hex encoded userID is correct
+	if hexUserID != messageUserID || len(messageUserID) != common.MessageUserIDLen {
+		w.Logger().Error(fmt.Errorf("submitted message (%s) is not in the correct format", message).Error())
+		return errors.New("userID from message does not match userID from request")
+	}
+
+	// Check if the signature is valid
```
Comment (tudor-malene) [inline tools/walletextension/wallet_extension.go:201]: where is the signature checked?

**Review by tudor-malene — CHANGES_REQUESTED** [review]
one question

File: tools/walletextension/wallet_extension.go (hunk 30 above)
Comment (zkokelj) [reply tools/walletextension/wallet_extension.go:201]: I checked that a few lines below.. I check if it was signed with the same address as is mentioned in the message
`if addressFromSignature != addressFromMessage {`

Comment (zkokelj) [reply tools/walletextension/wallet_extension.go:201]: Added ecdsa.Verify (first three things were already in the code)

File: tools/walletextension/api/routes.go [hunk 31]
```diff
@@ -118,6 +134,8 @@ func wsRequestHandler(walletExt *walletextension.WalletExtension, resp http.Resp
 func ethRequestHandler(walletExt *walletextension.WalletExtension, conn userconn.UserConn) {
 	body, err := conn.ReadRequest()
 	if err != nil {
+		conn.HandleError("Error: bad request")
+		walletExt.Logger().Error(fmt.Errorf("error reading request: %w", err).Error())
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:138]: Fixed.

File: tools/walletextension/api/routes.go [hunk 32]
```diff
@@ -160,6 +178,8 @@ func readyRequestHandler(_ *walletextension.WalletExtension, _ userconn.UserConn
 func generateViewingKeyRequestHandler(walletExt *walletextension.WalletExtension, conn userconn.UserConn) {
 	body, err := conn.ReadRequest()
 	if err != nil {
+		conn.HandleError("Error: bad request")
+		walletExt.Logger().Error(fmt.Errorf("error reading request: %w", err).Error())
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/api/routes.go:182]: fixed.

File: tools/walletextension/wallet_extension.go [hunk 33]
```diff
@@ -159,6 +162,165 @@ func (w *WalletExtension) SubmitViewingKey(address gethcommon.Address, signature
...
+}
+
+// check if message is in correct format and extracts userID and address from it
+func getUserIDAndAddressFromMessage(message string) (string, string, error) {
+	regex := regexp.MustCompile(common.MessageFormatRegex)
+	if regex.MatchString(message) {
+		params := regex.FindStringSubmatch(message)
+		return params[1], params[2], nil
+	}
+	return "", "", errors.New("invalid message format")
```
[the remark this answers is not in the document]
Comment (zkokelj) [reply tools/walletextension/wallet_extension.go:306]: fixed.

**Review by tudor-malene — APPROVED** [review]
lgtm

**Review by tudor-malene — APPROVED** [review]
lgtm

**Pull request merged**